api/v1alpha1/shared_types.go

```diff
@@ -333,3 +339,22 @@ type KubernetesHorizontalPodAutoscalerSpec struct {
 	// +optional
 	Behavior *autoscalingv2.HorizontalPodAutoscalerBehavior `json:"behavior,omitempty"`
 }
+
+// MergeType defines the type of merge operation
+type MergeType string
+
+const (
+	// StrategicMerge indicates a strategic merge patch type
+	StrategicMerge MergeType = "StrategicMerge"
+	// JSONMerge indicates a JSON merge patch type
+	JSONMerge MergeType = "JSONMerge"
+)
+
+// KubernetesMergeSpec defines how to perform the merge operation
+type KubernetesMergeSpec struct {
+	// Type is the type of merge operation to perform
+	Type MergeType `json:"type"`
+
+	// Object contains the raw configuration for merged object
```

Comment:
thoughts on creating a unique struct per type -
* KubernetesDeploymentPatchSpec
* KubernetesServicePatchSpec
which can define a more well defined `spec` field ?

Reply:
I think the patch spec should be dynamic like envoypatchpolicy jsonpatch value

Reply:
If we define more specific fields, aren`t we using the previous way, then what is the value of dynamic patch? I did not think of what common field is more needed in spec, can you elaborate ?

Reply:
my recommendation is to use a stricter type such as https://pkg.go.dev/k8s.io/api/apps/v1#DeploymentSpec
this is a little different that our field https://github.com/envoyproxy/gateway/blob/316a0a215ca4c040697bc0231f2ded9dabf59909/api/v1alpha1/shared_types.go#L54 since we are explicit about fields being set, reg `patch`, we cannot garuntee same o/p across releases since internally we may change implementation i.e. append more labels etc

Reply:
the advantage of using a strict type here is that patch semantic errors are surfaced to the user before even patching the two configs together

Reply:
I took another peek into this, its not possible to use the actual Deployment or Service type here since those CRD validations will kick in, which are not valid when its a patch
resolving this one